# Домашнє завдання до теми "Основи роботи з даними у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [12]:
import pandas as pd

In [3]:
import os
os.listdir()

['hw_3_2_Основи_роботи_з_даними_у_Pandas.ipynb',
 'Lecture_3_1_3_3_Pandas_and_numpy.ipynb',
 '.ipynb_checkpoints',
 'hw_3_1_Введення_у_Pandas_та_NumPy_.ipynb']

In [6]:
os.listdir('../data/')

['.DS_Store',
 'health_insurance_cross_sell_prediction.csv',
 'supermarket_sales.csv']

In [16]:
data_path = '../data/supermarket_sales.csv'
df = pd.read_csv(data_path)
df

Invoice ID Branch       City Customer type  Gender  \
0    750-67-8428      A     Yangon        Member  Female   
1    226-31-3081      C  Naypyitaw        Normal  Female   
2    631-41-3108      A     Yangon        Normal    Male   
3    123-19-1176      A     Yangon        Member    Male   
4    373-73-7910      A     Yangon        Normal    Male   
..           ...    ...        ...           ...     ...   
995  233-67-5758      C  Naypyitaw        Normal    Male   
996  303-96-2227      B   Mandalay        Normal  Female   
997  727-02-1313      A     Yangon        Member    Male   
998  347-56-2442      A     Yangon        Normal    Male   
999  849-09-3807      A     Yangon        Member  Female   

               Product line  Unit price  Quantity   Tax 5%      Total  \
0         Health and beauty       74.69         7  26.1415   548.9715   
1    Electronic accessories       15.28         5   3.8200    80.2200   
2        Home and lifestyle       46.33         7  16.2155   340.5255   
3         Health and beauty       58.22         8  23.2880   489.0480   
4         Sports and travel       86.31         7  30.2085   634.3785   
..                      ...         ...       ...      ...        ...   
995       Health and beauty       40.35         1   2.0175    42.3675   
996      Home and lifestyle       97.38        10  48.6900  1022.4900   
997      Food and beverages       31.84         1   1.5920    33.4320   
998      Home and lifestyle       65.82         1   3.2910    69.1110   
999     Fashion accessories       88.34         7  30.9190   649.2990   

          Date   Time      Payment    cogs  gross margin percentage  \
0     1/5/2019  13:08      Ewallet  522.83                 4.761905   
1     3/8/2019  10:29         Cash   76.40                 4.761905   
2     3/3/2019  13:23  Credit card  324.31                 4.761905   
3    1/27/2019  20:33      Ewallet  465.76                 4.761905   
4     2/8/2019  10:37      Ewallet  604.17                 4.761905   
..         ...    ...          ...     ...                      ...   
995  1/29/2019  13:46      Ewallet   40.35                 4.761905   
996   3/2/2019  17:16      Ewallet  973.80                 4.761905   
997   2/9/2019  13:22         Cash   31.84                 4.761905   
998  2/22/2019  15:33         Cash   65.82                 4.761905   
999  2/18/2019  13:28         Cash  618.38                 4.761905   

     gross income  Rating  
0         26.1415     9.1  
1          3.8200     9.6  
2         16.2155     7.4  
3         23.2880     8.4  
4         30.2085     5.3  
..            ...     ...  
995        2.0175     6.2  
996       48.6900     4.4  
997        1.5920     7.7  
998        3.2910     4.1  
999       30.9190     6.6  

[1000 rows x 17 columns]

1. За допомогою бібліотеки Pandas виведіть перші 4 і останні 4 записи у наборі даних (це дві окремі команди). Треба саме 4! Не 5.

In [18]:
df[:4]

Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.2155     7.4  
3  20:33      Ewallet  465.76                 4.761905       23.2880     8.4

In [20]:
df[-4:]  #або df.tail(4)

Invoice ID Branch      City Customer type  Gender         Product line  \
996  303-96-2227      B  Mandalay        Normal  Female   Home and lifestyle   
997  727-02-1313      A    Yangon        Member    Male   Food and beverages   
998  347-56-2442      A    Yangon        Normal    Male   Home and lifestyle   
999  849-09-3807      A    Yangon        Member  Female  Fashion accessories   

     Unit price  Quantity  Tax 5%     Total       Date   Time  Payment  \
996       97.38        10  48.690  1022.490   3/2/2019  17:16  Ewallet   
997       31.84         1   1.592    33.432   2/9/2019  13:22     Cash   
998       65.82         1   3.291    69.111  2/22/2019  15:33     Cash   
999       88.34         7  30.919   649.299  2/18/2019  13:28     Cash   

       cogs  gross margin percentage  gross income  Rating  
996  973.80                 4.761905        48.690     4.4  
997   31.84                 4.761905         1.592     7.7  
998   65.82                 4.761905         3.291     4.1  
999  618.38                 4.761905        30.919     6.6

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1F5XlzBOOTMsedpsYqz6NnXjIPTDe6jpx)

2. Виведіть рядки даних з індексами з 615 по 619 включно.

In [22]:
df[615:620]

Invoice ID Branch       City Customer type  Gender  \
615  509-29-3912      A     Yangon        Member  Female   
616  327-40-9673      B   Mandalay        Member    Male   
617  840-19-2096      C  Naypyitaw        Member    Male   
618  828-46-6863      A     Yangon        Member    Male   
619  641-96-3695      C  Naypyitaw        Member  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
615       Sports and travel       38.72         9  17.4240  365.9040   
616       Sports and travel       72.60         6  21.7800  457.3800   
617  Electronic accessories       87.91         5  21.9775  461.5275   
618      Food and beverages       98.53         6  29.5590  620.7390   
619     Fashion accessories       43.46         6  13.0380  273.7980   

          Date   Time      Payment    cogs  gross margin percentage  \
615  3/20/2019  12:24      Ewallet  348.48                 4.761905   
616  1/13/2019  19:51         Cash  435.60                 4.761905   
617  3/14/2019  18:10      Ewallet  439.55                 4.761905   
618  1/23/2019  11:22  Credit card  591.18                 4.761905   
619   2/7/2019  17:55      Ewallet  260.76                 4.761905   

     gross income  Rating  
615       17.4240     4.2  
616       21.7800     6.9  
617       21.9775     4.4  
618       29.5590     4.0  
619       13.0380     8.5

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1Otufe_TXRmjZxWigzBiVBLCAVUfEReXM)

3. Виведіть назви колонок у наборі даних разом з їхніми типами. Це можна зробити кількома способами, важливо аби в результаті були назви колонок і відповідні їм типи.

In [34]:
df.dtypes


Invoice ID                  object
Branch                      object
City                        object
Customer type               object
Gender                      object
Product line                object
Unit price                 float64
Quantity                     int64
Tax 5%                     float64
Total                      float64
Date                        object
Time                        object
Payment                     object
cogs                       float64
gross margin percentage    float64
gross income               float64
Rating                     float64
dtype: object

Очікуваний результат (один з можливих форматів):

![](https://drive.google.com/uc?export=view&id=1ASh6-S-_f3X9fTQAPIvnHMFZamTBjNkL)

4. Користуючись методом `DataFrame.describe()` виявіть 1, 25, 50, 75 і 99 персентилі колонки `gross income`. Знайдіть, яке значення відповідає 75отму персентилю цієї колонки і що це означає? Як би Ви його пояснили простими словами колегам? Відповідь напишіть текстом в окремій клітинці типу Markdown.

In [40]:
display(df['gross income'].describe (percentiles=[0.01,0.25,0.5,0.75, 0.99]))

count    1000.000000
mean       15.379369
std        11.708825
min         0.508500
1%          0.913735
25%         5.924875
50%        12.088000
75%        22.445250
99%        45.250750
max        49.650000
Name: gross income, dtype: float64

Очікуваний результат:

Валовий дохід кожного з 75% всіх покупців  менші або рівні 22.445250 платіжних одиниць, а 25% — більші.

![](https://drive.google.com/uc?export=view&id=1bDa1YFUWDsPm78Lnpy2geQND8M1HEQiT)

5. Виведіть кількість інвойсів за кожним унікальним містом в наборі даних.

  Підказка: інвойс - це один рядок в цьому наборі даних, ми можемо цим скористатись аби швидко отримати відповідь.

In [50]:
df.City.value_counts()

City
Yangon       340
Mandalay     332
Naypyitaw    328
Name: count, dtype: int64

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1G8c6M875NOW7oU3OtZegTOnyD2y8lhAV)

6. Виведіть, який відсоток інвойсів припадає на кожне унікальне місто.

In [52]:
df.City.value_counts(normalize = True)

City
Yangon       0.340
Mandalay     0.332
Naypyitaw    0.328
Name: proportion, dtype: float64

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1yZZpWhL2y8DX9rSRioYCRh1VbJgcf3fY)

7.1. Використовуючи фільтрацію даних, знайдіть середнє значення `gross income` для міст `Yangon` та `Mandalay`. В якому місті воно більше?

7.2 Сформулюйте 3 гіпотези, чому в одному місті gross income вище, як можна дізнатись чи правдива ця гіпотеза з використанням даних (наявних і додаткових), та як висновки з цих гіпотез можна використати для прийняття рішень як підвищити прибуток в мережі супермаркетів. Відповідь напишіть текстом в окремій клітинці типу Markdown.

Це завдання - на продуктове мислення. Якщо не знаєте, які гіпотези сформулювати - напишіть ті, які здаються вам зовсім примітивними. Важливо навчитись висловлювати свої гіпотези щодо даних, і з часом та досвідом вони будуть ставати щоразу глибше і точніше. Весь аналіз даних будується на тому, що ми перевіряємо певні гіпотези та робимо висновки і рекоемндації на основі своїх досліджень.

In [56]:
yangon_data = df[df['City'] == 'Yangon']


In [68]:
avg_yangon = yangon_data['gross income'].mean()
avg_yangon

14.874001470588235

In [62]:
mandalay_data = df[df['City'] == 'Mandalay']

In [70]:
avg_mandalay = mandalay_data['gross income'].mean()
avg_mandalay 

15.232024096385542

In [66]:
if avg_yangon > avg_mandalay:
    city = 'Yangon'
    print(f'Досліджуючи середнє значення gross income для міст Yangon та Mandalay, ми дійшли висновку що у місті {city} середній gross income вищий')
elif avg_yangon < avg_mandalay:
    city = 'Mandalay'
    print(f'Досліджуючи середнє значення gross income для міст Yangon та Mandalay, ми дійшли висновку що у місті {city} середній gross income вищий')
else:
    print(f'Досліджуючи середнє значення gross income для міст Yangon та Mandalay, ми дійшли висновку що у цих містах середній gross income на однаковому рівні')

Досліджуючи середнє значення gross income для міст Yangon та Mandalay, ми дійшли висновку що у місті Mandalay середній gross income вищий


1. Гіпотеза
   Жінки купують більше, ніж чоловіки.
   

In [78]:
# Перевірка, чи жінок більше у Mandalay, ніж у Yangon
count_female_yangon = yangon_data[yangon_data['Gender'] == 'Female'].shape[0]
count_female_yangon


161

In [94]:
count_male_yangon = yangon_data[yangon_data['Gender'] == 'Male'].shape[0]
count_male_yangon

179

In [76]:
count_female_mandalay = mandalay_data[mandalay_data['Gender'] == 'Female'].shape[0]
count_female_mandalay

162

In [96]:
count_male_mandalay = mandalay_data[mandalay_data['Gender'] == 'Male'].shape[0]
count_male_mandalay

170

In [82]:
# Підрахунок середнього значення gross income за гендером у Yangon
avg_yangon_female = yangon_data[yangon_data['Gender'] == 'Female']['gross income'].mean()
avg_yangon_female



15.755447204968947

In [84]:
avg_yangon_male = yangon_data[yangon_data['Gender'] == 'Male']['gross income'].mean()
avg_yangon_male

14.081192737430168

In [88]:
# Підрахунок середнього значення gross income за гендером у Mandalay
avg_mandalay_female = mandalay_data[mandalay_data['Gender'] == 'Female']['gross income'].mean()
avg_mandalay_female 

15.557993827160495

In [90]:
avg_mandalay_male = mandalay_data[mandalay_data['Gender'] == 'Male']['gross income'].mean()
avg_mandalay_male

14.921394117647058

In [98]:
if avg_mandalay_female > avg_mandalay_male:
    print("У Mandalay жінки витрачають більше, ніж чоловіки.")
else:
    print("У Mandalay чоловіки витрачають більше, ніж жінки.")


У Mandalay жінки витрачають більше, ніж чоловіки.


In [100]:
if count_female_mandalay > count_female_yangon:
    print("У Mandalay більше жінок, ніж у Yangon.")
else:
    print("У Yangon більше жінок, ніж у Mandalay.")

У Mandalay більше жінок, ніж у Yangon.


Проміжний висновок по Гіпотезі 1 - має право на життя, треба досліджувати глибше. якщо підтвердиться використати маркетингові стратегіі для залучення та утримання покупців-жінок.

In [ ]:
2. Гіпотеза
  Є пряма залежність між продуктовою лінійкою та валовим доходом.

In [108]:
# Дані для Yangon
yangon_product_share = df[df['City'] == 'Yangon']['Product line'].value_counts(normalize=True) * 100

# Дані для Mandalay
mandalay_product_share = df[df['City'] == 'Mandalay']['Product line'].value_counts(normalize=True) * 100

# Виведемо результати
print("Частка продуктових лінійок у Yangon:")
print(yangon_product_share)

print("\nЧастка продуктових лінійок у Mandalay:")
print(mandalay_product_share)

Частка продуктових лінійок у Yangon:
Product line
Home and lifestyle        19.117647
Electronic accessories    17.647059
Sports and travel         17.352941
Food and beverages        17.058824
Fashion accessories       15.000000
Health and beauty         13.823529
Name: proportion, dtype: float64

Частка продуктових лінійок у Mandalay:
Product line
Fashion accessories       18.674699
Sports and travel         18.674699
Electronic accessories    16.566265
Health and beauty         15.963855
Food and beverages        15.060241
Home and lifestyle        15.060241
Name: proportion, dtype: float64


Проміжне спостереження по Гіпотеза 2: треба підсилити у Yangon продуктову лінійку "Fashion accessories" та "Sports and travel". Та проаналізувати через квартал як змінилась валовий прибуток у Yangon.

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1JdXUmK46WcoXUJYP7NS2SJlr2XjBbpQ8)

8. Для продовження дослідження різниці у прибутках в магазинах давайте зʼясуємо, чи відрізняється маржа ('gross margin percentage') для різних інвойсів в магазині? Перевірте це програмно на даних виходячи з того функціоналу, який ми вже пройшли у лекціях.

Дайте на основі отриманого результату відповідь: "У всіх всіх інвойсах однакова маржа" або "Маржа відрізняється для різних інвойсів".

In [111]:
df.Payment.value_counts(normalize = True)

Payment
Ewallet        0.345
Cash           0.344
Credit card    0.311
Name: proportion, dtype: float64

In [114]:
ewallet_data = df[df['Payment'] == 'Ewallet']
cash_data = df[df['Payment'] == 'Cash']
credit_card_data = df[df['Payment'] == 'Credit card']

# Обчислення середньої маржі для кожного типу оплати
avg_ewallet_margin = ewallet_data['gross margin percentage'].mean()
avg_cash_margin = cash_data['gross margin percentage'].mean()
avg_credit_card_margin = credit_card_data['gross margin percentage'].mean()

# Виведення результатів
print("Середня маржа для типу оплати Ewallet:", avg_ewallet_margin)
print("Середня маржа для типу оплати Cash:", avg_cash_margin)
print("Середня маржа для типу оплати Credit Card:", avg_credit_card_margin)

Середня маржа для типу оплати Ewallet: 4.761904762
Середня маржа для типу оплати Cash: 4.761904762
Середня маржа для типу оплати Credit Card: 4.7619047619999995


Очікуваний результат: відповідь щодо маржі в усіх інвойсах на основі отриманих результатів

In [ ]:
Середня маржа для  різних типів оплат практично однакова.

9. Врахуовуючи висновок в попередньому завданні, дослідимо як продаються в цих містах різні категорії товарів. Виведіть відсотковий розпоіділ кількості інвойсів по категоріям товарів `Product line` в містах `Yangon` та `Mandalay`.

Див. Гіпотезу 2

Очікуваний результат для кожного з міст:

![](https://drive.google.com/uc?export=view&id=1IbG58YK5KVqC-wCnllF1btgEgfWTbi_O)

![](https://drive.google.com/uc?export=view&id=1aKOkLmlMq1sOh4GbwyaQyjfhNJzdU4Yh)